### PR TITLE
a borrow market

### PR DESCRIPTION
**Pull request by wakamex**
add a borrow market that allows agents to open and close borrows against a variable rate

### CODE REVIEW

File: elfpy/markets/borrow.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+        setattr(self, key, value)
+
+
+@types.freezable(frozen=False, no_new_attribs=False)
+@dataclass
+class MarketState(base_market.BaseMarketState):
+    r"""The state of an AMM
+
+    Implements a class for all that that an AMM smart contract would hold or would have access to
+    For example, reserve numbers are local state variables of the AMM.  The vault_apr will most
```
Comment (dpaiton) [inline elfpy/markets/borrow.py:64]: we don't have a vault_apr correct?

File: elfpy/markets/borrow.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+
+import elfpy.types as types
+
+# TODO: for now...
+# pylint: disable=duplicate-code
+
+
+class MarketActionType(Enum):
+    r"""
+    Enumerate actions available in this market
```
Comment (dpaiton) [inline elfpy/markets/borrow.py:21]: make one line

File: elfpy/markets/borrow.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    def deposit_amount(self) -> dict[types.TokenType, float]:
+        """The amount of deposited asset in the market"""
+        return {key: value * self.deposit_spot_price[key] for key, value in self.collateral.items()}
+
+    @property
+    def total_market_profit(self) -> float:
+        """
+        From the market's perspective, the profit is the difference between the borrowed and deposited assets
+        """
+        # how do we calcualte this?
```
Comment (dpaiton) [inline elfpy/markets/borrow.py:129]: are all of these comments still needed?

File: elfpy/markets/borrow.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+    def check_market_updates(self, market_deltas: MarketDeltas) -> None:
+        """Check market update values to make sure they are valid"""
+        for key, value in market_deltas.__dict__.items():
+            if value:  # check that it's instantiated and non-empty
+                assert np.isfinite(value), f"markets.update_market: ERROR: market delta key {key} is not finite."
+
+    def log_market_step_string(self) -> None:
+        """Logs the current market step"""
+        # TODO: This is a HACK to prevent test_sim from failing on market shutdown
```
Comment (dpaiton) [inline elfpy/markets/borrow.py:325]: can remove this TODO

**Review by dpaiton — APPROVED** [review]
woo!

**Review by wakamex — COMMENTED** [review]
lgtm

**Pull request merged**